docs/proposals/external-ca.md

```diff
@@ -0,0 +1,112 @@
+
+```yaml
+type: Mesh
+name: default
+mtls:
+  ca:
+    external: {}
+  enabled: true
+```
+
+## How to provide certificates to Kuma
+
+There are several ways we can implement providing root certs for Kuma
+
+### File
+
+```yaml
+type: Mesh
+name: default
+mtls:
+  ca:
+    external:
+      keyPath: /path/to/file
+      certPath: /path/to/file
+  enabled: true
+```
+
+It seems simple, but storing key files on the machines with control plane is not ideal.
+The bigger problem though is that you have to synchronize files between CP instances. Once it's not in sync (Puppet fails, any other reason), the mesh will break.
+We already have storage to share between CP instances and it's Postgres on Universal and Etcd on K8S.
+  
+### HTTP Server + kumactl
+
+We can adopt the same pattern as with Dataplane Token Server.
+The server is by default exposed only on `localhost`. Optionally user can expose server to public by generating cert+key for server and client certificates.
+
+User can upload a certificate with key in two ways.
+
+1) HTTP API
+```bash
+$ cat request.json
+{
+  "key": "-----BEGIN RSA PRIVATE KEY-----\nMLLLEpSDFAGKC24gArqiy1c3pFT3FSk5FE51A4ALAadeR...",
+  "cert": "-----BEGIN CERTIFICATE-----\nZZXDFjC3An4gAwFBAFIJ4FFZ66emAA3AZA0GZSqGSGHt...",
+  "mesh": "demo"
+}
+$ curl -XPUT localhost:5691/certificates --data @request.json
+```
+
+2) kumactl
+```bash
+$ cat key.pem
+-----BEGIN RSA PRIVATE KEY-----\nMLLLEpSDFAGKC24gArqiy1c3pFT3FSk5FE51A4ALAadeR...
+$ cat cert.pem
+-----BEGIN CERTIFICATE-----\nZZXDFjC3An4gAwFBAFIJ4FFZ66emAA3AZA0GZSqGSGHt...
+$ kumactl upload-certificate --mesh demo --key-file key.pem --cert-file cert.pem
+```
+
+This can be either yet another HTTP server in Kuma or we can embed this into Dataplane Token Server and rename the server.
+
+### kuma-cp sub command
+
+Right now, you can run the CP with `kuma-cp run`.
+We can introduce new command:
+
+```bash
+$ cat key.pem
+-----BEGIN RSA PRIVATE KEY-----\nMLLLEpSDFAGKC24gArqiy1c3pFT3FSk5FE51A4ALAadeR...
+$ cat cert.pem
+-----BEGIN CERTIFICATE-----\nZZXDFjC3An4gAwFBAFIJ4FFZ66emAA3AZA0GZSqGSGHt...
+$ kuma-cp upload-certificate --mesh demo --key-file key.pem --cert-file cert.pem
+```
+
+## Flow
+
+Either with `HTTP Server + kumactl` or `kuma-cp sub command` we should first upload the certificate.
+Then when we apply a mesh with `enabled: true` and `ca: external`, we should validate if certificate is present.
+
+## Certificate modification
+
+Having only builtin CA without ways to modify it, we did not support certificate changes.
+Once you change the certificate, current dataplanes will have certs signed by old root cert and new dataplanes will receive certs signed by new root cert, therefore they won't be able to communicate.
+
+Eventually we should introduce the support for certificate rotation, so user can revoke certificate or migrate CA without downtime, but this is a case for another proposal.
+For now we can adopt simpler approach:
+
+1. Block update operation
```

Comment:
We should not allow changing CA certificate until we add support for CA certificate rotation.

If necessary, a user can always fallback to a sequence of steps
```
disable mTLS => add a new CA certificate => enable mTLS
```

Reply:
Ok, this flow makes more sense